File: java/heap_sort.java
import java.util.*;

public class heap_sort{	
    private static void sort(int arr[]){
        int n = arr.length;
 
      	// Build max heap
        for (int i = n / 2 - 1; i >= 0; i--) {
          heapify(arr, n, i);
        }
            
		// Heap sort
        for (int i=n-1; i>=0; i--){
            int temp = arr[0];
            arr[0] = arr[i];
            arr[i] = temp;
 
          	// Heapify root element
            heapify(arr, i, 0);
        }
    }
 
    private static void heapify(int arr[], int n, int i){
      	// Find largest among root, left child and right child
        int largest = i; 
        int l = 2*i + 1; 
        int r = 2*i + 2;  
 
        if (l < n && arr[l] > arr[largest])
            largest = l;
 
        if (r < n && arr[r] > arr[largest])
            largest = r;
 
      	// Swap and continue heapifying if root is not largest
        if (largest != i){
            int swap = arr[i];
            arr[i] = arr[largest];
            arr[largest] = swap;
 
            heapify(arr, n, largest);
        }
    }
 
    public static void main(String args[]){
        int arr[] = {1,12,9,5,6,10};
        sort(arr);
        System.out.println(Arrays.toString(arr));
    }
}

File: c/heapsort.c
#include<stdio.h> int left(int i)
{
return 2*i+1;
}
int right(int i)
{
return 2*i+2;
}
void max_heapify(int *,int,int); void build_maxheap(int *a,int n)
{
int i;
for(i=(n-1)/2;i>=0;i--){ max_heapify(a,i,n);
}
}
void max_heapify(int *a,int i,int n) {
int largest,temp;
int l=left(i);
int r=right(i);
if(l<=n-1 && a[l]>a[i]){

largest=l;
}
else{
largest=i;
}
if(r<=n-1 && a[r]>a[largest]){ largest=r;
}
if(largest!=i){
temp=a[i];
a[i]=a[largest];
a[largest]=temp; max_heapify(a,largest,n);
}
}
int main()
{
int a[100],i,k,n,temp;
printf(“Enter number of elements”); scanf(“%d”,&n);
printf(“Enter elements “); for(i=0;i<n;i++){
scanf(“%d”,&a[i]);
}
build_maxheap(a,n); for(i=n-1;i>0;i--)
{
temp=a[0];
a[0]=a[i];
a[i]=temp; printf(“Pass %d: “,n-i); for(k=0;k<n;k++) printf(“%d “,a[k]); printf(“\n”); max_heapify(a,0,i);
}
printf(“Sorted array “); for(i=0;i<n;i++){ printf(“%d “,a[i]);
}
return 0;
}
